### DISCUSSION TITLE
An async hot-standby from batched state transfer

### DISCUSSION THREAD

**Discussion by rhenwood-nasa** [quay/quay#2463] [Q&A]
Hi,

I'm looking to deploy Project Quay across two siloed environments. Silo_A will have a full-featured Project Quay - web gui, image pull and push. Silo_B will have a read-only Project Quay - and will not need the web gui, only image pull. Silo_A and Silo_B are connected by a gateway that only permits blob transfer from Silo_A to Silo_B. The connection between Silo_A and Silo_B is high bandwidth and high latency (one transfer every five minutes). The connection between Silo_B and Silo_A is very low bandwidth and even higher latency.

My current design idea is to generate a blob after Silo_A changes state. The blob will consist of the changes. The blob will be sent to Silo_B. Silo_B will integrate the changes and the result will be a Silo_B has the same registry contents as Silo_A. This sound like this setup might be described as 'an asynchronous hot-standby'. Or maybe off-site hot backup?

I think there are three components of Project Quay that I need to care about their state.

1. PostgreSQL. Used to record image metadata, permissions. This must be sent. Pg has log shipping that seems be appropriate for my design.
2. Redis. It seems this is used for highly transient data - user sessions? It looks like there is log shipping available.
3. S3 image storage. The images, must be sent. A pretty well understood problem.

Questions:
 1. Is my characterization of the Project Quay's use of PostgreSQL, Redis, S3 storage accurate?
 2. Have I missed any components of Project Quay where the state matters in my use case?
 3. I care about who pulls what from Silo_B Quay but can I ignore transferring Redis state from Silo_A?
 4. Any other thoughts folks have?

**Comment by michaelalang** [comment]
@rhenwood-nasa just to ensure you get the picture I have in mind...

PostgreSQL and S3 are the mandatory part to sync between sites.

Redis isn't as it keeps internal and build as well as reports/logs. 
None of those items is available in sync state on the replica site. Your builders most-likely will need to be local to the instance as well as your reports only make sense once you switch the standby/replica site primary.

I have been showcasing `read-only` replicas (n^) with independent PostgreSQL and S3 buckets to the primary Quay instance. 
The concept provides retry-failure-error scenarios (your pull/push will always fail if in the middle Quay fails/fails-over) and RPO depends on how much latency/sync-period you can take (WAL/sec and S3/sec might be feasible on Prem but are really expensive in the Cloud)

The next you need to understand is that, even though pull access will be available 100% (retry-aware) push requires a restart of the PostgreSQL and the Quay instance into the proper mode.

Hope that clarifies some of the questions already ...

**Comment by rhenwood-nasa** [comment]
Hi @michaelalang,

Thanks for your prompt reply and sharing your experience - it is very valuable for me. I didn't mention in my initial comment that I reviewed the replica feature within Project Quay and judged it as unsuitable in my case as I cannot make a direct connection between Silo_A and Silo_B. I would be happy to learn that there was an async replica feature in Project Quay? In the absence of that, I was planning to manually send the PostgreSQL and S3 state as I've described. Can you share a link to your read-only showcase? I think I would find that useful to review.

What I now think I understand is:

- I can ignore Redis in my case (yay!)
- I can send a PostgreSQL WAL and S3 blobs and expect to achieve the behavior I am looking for. They can be added to the read-only (Silo_B) instance of Quay - BUT if I want to push to Silo_B Quay, I need to restart both PostgreSQL and Quay. I do not intend to Push to Silo_B, so I think I'm not covered by this requirement.

Note: I briefly considered setting up two Quays within Silo_A, configuring the 'replica' feature between them, passing the connection between the Quays through a 'recording proxy' and recording of the traffic, and then replying it into Silo_B. I will keep that in reserve if for some reason my simpler plan is unsuitable.

I'm beginning to put together a POC, so I will update this thread with my findings.

**Comment by michaelalang** [comment]
here's my experience in that sense ... 
[Red Hat Quay (3.8.10) independent registries (1).pdf](https://github.com/quay/quay/files/13284834/Red.Hat.Quay.3.8.10.independent.registries.1.pdf)
